### PR TITLE
Multipart File Upload and useGETForQueries support

### LINKED ISSUE #57: Bring Links up to par with what package:graphql has
- [x] Example of Auth link functionality
- [x] Error Link
- [x] Websocket Link
- [x] File upload
- [x] http_link should have something like the useGETForQueries option as well
- [ ] persisted queries were recently contributed

### PR DESCRIPTION
**Pull request by micimize**
This extracts the logic for multipart file upload and `useGETForQueries` from `graphql/client.dart`, with accompanying tests

**Major changes**:
* `useGetForQueries` argument to `HttpLink`
* using `http.Client.send` instead of `post`
* `HttpLink` stores `uri` as a `Uri`
* I reworked `HttpLink._serializeRequest` into an exception handling HOF to handle the different paths, but really I think `RequestSerializer` should have a generic `Serialized` type which is `http.BaseRequest` for `HttpLink`. Otherwise it can't really handle the serialization concern IMO

This addresses some of #57

### CODE REVIEW

**Comment by micimize** [comment]
I also considered adding this optional termination logic, but I guess it's possible for servers to handle subscriptions? 
```dart
    if (request.isSubscription) {
      if (forward == null) {
        throw RequestFormatException(
          originalException: AssertionError(
            "HttpLink is the final link, but cannot make subscription requests itself.",
          ),
          request: request,
        );
      }
      yield* forward(request);
    }
```

`Link.split((request) => request.isSubscription, websocketLink, link);` works alright, but I'm a bit worried about usability, as initially I did `link.concat(websocketLink)` and got confused by the serverside errors

File: links/gql_http_link/test/gql_http_link_test.dart [hunk 1]
```diff
@@ -132,30 +112,15 @@ void main() {
...
-          any,
-          body: anyNamed("body"),
-          headers: {
-            "Content-type": "application/json",
-            "Accept": "*/*",
-          },
-          encoding: anyNamed("encoding"),
-        ),
-      ).called(1);
+      verify(client.send(any)).called(1);
```
Comment (klavs) [inline links/gql_http_link/test/gql_http_link_test.dart:115]: Seems to me that this is not testing what it says it does.

File: links/gql_http_link/test/multipart_upload_test.dart [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    setUp(() {
+      mockHttpClient = MockClient();
+
+      httpLink = HttpLink(
+        "http://localhost:3001/graphql",
+        httpClient: mockHttpClient,
+      );
+    });
+
+    test("upload success", () async {
```
Comment (klavs) [inline links/gql_http_link/test/multipart_upload_test.dart:43]: Please split this test up in multiple tests.

**Comment by micimize** [comment]
@klavs right, idk why I thought testing `.send(any)` over and over again made sense 😅 

* recovered the old tests with `isA().having` statements
* added a test group for `useGETForQueries`
* broke multipart tests into a one for `request`/`response`
* and made the multipart upload tests more intelligible (IMO)

File: links/gql_http_link/lib/src/link.dart [hunk 3]
```diff
@@ -162,19 +154,74 @@ class HttpLink extends Link {
...
+    final headers = {
+      "Content-type": "application/json",
+      "Accept": "*/*",
+      ...defaultHeaders,
+      ...contextHeaders,
+    };
+
+    final fileMap = extractFlattenedFileMap(body);
+
+    final method = (fileMap.isEmpty && useGETForQueries && request.isQuery)
```
[the remark this answers is not in the document]
Comment (klavs) [reply links/gql_http_link/lib/src/link.dart:173]: `method` is never used as String. You can use some boolean instead.

**Comment by klavs** [comment]
Please update the changelog as well.

**Comment by micimize** [comment]
* `final useGetForThisRequest = fileMap.isEmpty && useGETForQueries && request.isQuery;`
* added `useGETForQueries + gqlQueryWithFiles` test to upload
* Turned a todo into #117

**Comment by micimize** [comment]
idk why but multipack is failing:
```dart
Failed to precompile multipack:multipack:
/github/home/.pub-cache/hosted/pub.dartlang.org/multipack-0.1.2/bin/multipack.dart:11:33: Error: 'sortedTopologicalOrdering' isn't a function or method and can't be invoked.
      .sortedTopologicalOrdering()
                                ^^^...
##[error]Process completed with exit code 1.
```

**Comment by klavs** [comment]
Looks like a breaking minor change in https://pub.dev/packages/directed_graph

Fixing multipack...

**Comment by klavs** [comment]
multipack is now fixed.

It seems that you must run `multipack pubspec clean`

**Comment by micimize** [comment]
It isn't seeing any of the packages
```bash
pub global list | grep multipack
multipack 0.1.3
multipack --only gql_http_link pubspec clean
Unhandled exception:
"gql_http_link" is not an allowed value for option "only".

Usage: multipack <command> [arguments]

Global options:
-h, --help              Print this usage information.
-d, --since=<commit>    Consider only the changed packages.
-o, --only              Whitelist packages, skipping those not included for this command.
                        []
-s, --skip              Blacklist packages for this command.
                        []

Available commands:
  analyze   Run analyzer.
  exec      Execute any command.
  fmt       Run formatter.
  pub       Run pub.
  pubspec   Update pubspec.yaml

Run "multipack help <command>" for more information about a command.
#0      CommandRunner.usageException (package:args/command_runner.dart:95:7)
#1      CommandRunner.parse (package:args/command_runner.dart:123:35)
#2      CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:36)
#3      new Future.sync (dart:async/future.dart:224:31)
#4      CommandRunner.run (package:args/command_runner.dart:112:14)
#5      main (file:///Users/mjr/.pub-cache/hosted/pub.dartlang.org/multipack-0.1.3/bin/multipack.dart:54:16)
<asynchronous suspension>
#6      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
#7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
```

**Comment by micimize** [comment]
@klavs figured it out – the CI does `git diff --exit-code pubspec.yaml` so bumping the version (and any other changes to `pubspec.yaml` will break CI. So, I reverted my version bump

**Comment by klavs** [comment]
@micimize, Could you please fix the conflict introduced by me merging the PRs in the wrong order?

**Comment by micimize** [comment]
@klavs should be good to go now!

**Pull request merged**